File: medium/problem2/__main__.py
"""
Runtime: 64 ms, faster than 92.01% of Python3 online submissions for Add Two Numbers.
Memory Usage: 12.8 MB, less than 100.00% of Python3 online submissions for Add Two Numbers.
17/12/2019
DD/MM/YY
"""


class ListNode:
    """
    Definition for singly-linked list.
    Ommit this block in Leetcode
    """

    def __init__(self, x):
        self.val = x
        self.next = None


class Solution:
    def addTwoNumbers(self, l1: ListNode, l2: ListNode) -> ListNode:

        if l1 == None or l1.val == 0 and l1.next == None:
            return l2
        if l2 == None or l2.val == 0 and l2.next == None:
            return l1

        values1, values2 = "", ""
        while l1:
            values1 += str(l1.val)
            l1 = l1.next

        while l2:
            values2 += str(l2.val)
            l2 = l2.next

        final = str(int(values1[::-1]) + int(values2[::-1]))[::-1]
        result = ListNode(0)

        res = ListNode(int(final[0]))
        result.next = res

        for i in final[1:]:
            res.next = ListNode(i)
            res = res.next

        return result.next

File: medium/problem2/main.go
package main

// This implementation uses the existing lists instead of creating a new one

/*
Runtime: 11 ms, beats 72% of Go submissions
Memory Usage: 4.3 MB, beats 100% of Go submissions
2023-02-02
*/

// ListNode Definition for singly-linked list.
type ListNode struct {
	Val  int
	Next *ListNode
}

func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
	var carry int
	l1Len := 0
	l2Len := 0
	curNode1 := l1
	curNode2 := l2
	var lastNode1, lastNode2 *ListNode

	for curNode1 != nil || curNode2 != nil {
		total := carry

		if curNode1 != nil {
			total += curNode1.Val
		}
		if curNode2 != nil {
			total += curNode2.Val
		}
		carry = total / 10
		total -= carry * 10

		if curNode1 != nil {
			curNode1.Val = total
			lastNode1 = curNode1
			curNode1 = curNode1.Next
			l1Len++
		}
		if curNode2 != nil {
			curNode2.Val = total
			lastNode2 = curNode2
			curNode2 = curNode2.Next
			l2Len++
		}
	}

	if l1Len > l2Len {
		if carry != 0 {
			newNode := &ListNode{carry, nil}
			lastNode1.Next = newNode
		}

		return l1
	}

	if carry != 0 {
		newNode := &ListNode{carry, nil}
		lastNode2.Next = newNode
	}

	return l2
}
